packages/grid/src/vaadin-grid.js

```diff
@@ -667,12 +638,74 @@ class Grid extends ElementMixin(
     }
 
     this.__intrinsicWidthCache = new Map();
+    // Cache the viewport rows to avoid unnecessary reflows while measuring the column widths
+    const fvi = this._firstVisibleIndex;
+    const lvi = this._lastVisibleIndex;
+    this.__viewportRowsCache = this._getVisibleRows().filter((row) => row.index >= fvi && row.index <= lvi);
+
+    // Pre-cache the intrinsic width of each column
+    this.__calculateAndCacheIntrinsicWidths(cols);
 
     cols.forEach((col) => {
       col.width = `${this.__getDistributedWidth(col)}px`;
     });
   }
 
+  /**
+   * Toggles the cell content for the given column to use or not use auto width.
+   *
+   * While content for all the column cells uses auto width (instead of the default 100%),
+   * their offsetWidth can be used to calculate the collective intrinsic width of the column.
+   *
+   * @private
+   */
+  __setVisibleCellContentAutoWidth(col, autoWidth) {
+    col._allCells
+      .filter((cell) => {
+        if (this.$.items.contains(cell)) {
+          return this.__viewportRowsCache.includes(cell.parentElement);
+        }
+        return true;
+      })
+      .forEach((cell) => {
+        cell.__measuringAutoWidth = autoWidth;
+        cell._content.style.width = autoWidth ? 'auto' : '';
+        cell._content.style.position = autoWidth ? 'absolute' : '';
+      });
+  }
+
+  /**
+   * Returns the maximum intrinsic width of the cell content in the given column.
+   * Only cells which are marked for measuring auto width are considered.
+   *
+   * @private
+   */
+  __getAutoWidthCellsMaxWidth(col) {
+    // Note: _allCells only contains cells which are currently rendered in DOM
+    return col._allCells.reduce((width, cell) => {
+      // Add 1px buffer to the offset width to avoid too narrow columns (sub-pixel rendering)
+      return cell.__measuringAutoWidth ? Math.max(width, cell._content.offsetWidth + 1) : width;
+    }, 0);
+  }
+
+  /**
+   * Calculates and caches the intrinsic width of each given column.
+   *
+   * @private
+   */
+  __calculateAndCacheIntrinsicWidths(cols) {
+    // Make all the columns use auto width at once before measuring to
+    // avoid reflows in between the measurements
+    cols.forEach((col) => this.__setVisibleCellContentAutoWidth(col, true));
```

Comment:
Is it like a proper replacement for

```js
col.width = 'auto';
col.flexGrow = 0;
```

?

Reply:
I'd expect so. The approach to modify the cell _content_ element directly also has some additional performance benefits:
- Changing `column.width = 'auto'...` affects the column cells on _all_ the rows, not just those we want to include for measuring (those inside the viewport). Modifying only the cells we need requires less relayouting.
- Changing the `column.width` (or `cell.style.width` directly) requires the whole row to layout, because the cell element is a direct child of the row (flexbox). Not all columns on a row are necessarily using auto-width so that would cause unnecessary overhead. Modifying the content inside a cell element doesn't affect anything outside the cell.